storage/watchable_store_test.go

```diff
@@ -222,3 +240,73 @@ func TestUnsafeAddWatching(t *testing.T) {
 		}
 	}
 }
+
+func TestMapWatchingToEvents(t *testing.T) {
+	var (
+		testN         = 5
+		testKey       = []byte("Foo")
+		testValue     = []byte("Bar")
+		testKeys      = make([][]byte, testN)
+		testValues    = make([][]byte, testN)
+		testWatchings = make([]*watching, testN)
+		testEvents    = make([]storagepb.Event, testN)
+	)
+	for i := 0; i < testN; i++ {
+		k, v := testKey, testValue
+		if i > 0 {
+			k = []byte(fmt.Sprintf("%s_%d", testKey, i))
+			v = []byte(fmt.Sprintf("%s_%d", testValue, i))
+		}
+		testKeys[i] = k
+		testValues[i] = v
+		testWatchings[i] = &watching{
+			key:    k,
+			prefix: true,
+		}
+		testEvents[i] = storagepb.Event{
+			Type: storagepb.PUT,
+			Kv: &storagepb.KeyValue{
+				Key:   testKey,
+				Value: testValue,
+			},
+		}
+	}
+
+	tests := []struct {
+		sync map[string]map[*watching]struct{}
+		evs  []storagepb.Event
+
+		wwe map[*watching][]storagepb.Event
+	}{
+		{
```

Comment:
Got it. Thanks!

On Thu, Dec 31, 2015 at 9:53 PM, Xiang Li <notifications@github.com> wrote:

> In storage/watchable_store_test.go
> <https://github.com/coreos/etcd/pull/4091#discussion_r48676262>:
>
> > +		testEvents[i] = storagepb.Event{
> > +			Type: storagepb.PUT,
> > +			Kv: &storagepb.KeyValue{
> > +				Key:   testKey,
> > +				Value: testValue,
> > +			},
> > +		}
> > +	}
> > +
> > +	tests := []struct {
> > +		sync map[string]map[*watching]struct{}
> > +		evs  []storagepb.Event
> > +
> > +		wwe map[*watching][]storagepb.Event
> > +	}{
> > +		{
>
> i hope to see:
>
>    1.
>
>    no watchings in the sync and some events in the evs array. we should
>    expect an empty wwe map.
>    2.
>
>    one watchings in the sync map and one event that does not match the
>    key of the watching. we should expect an empty wwe map
>    3.
>
>    one watching in the sync map and one event that matches the key of
>    that watching. we should expect a wwe map with that watching and the value
>    is the event.
>    4.
>
>    two watching in the sync map that watches two different keys and one
>    event that matches the key of one of the watching. we should expect a wwe
>    map with the matched watching and the value is the matched event.
>    5.
>
>    two watchings in the sync map that watches on the same key and one
>    event that matches the key. we should expect a wwe map with two watchings
>    and both of them have the value of the event.
>
> I think you can get the idea and come up more test cases. We need to
> carefully design the test case rather than make it random when we can.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/4091/files#r48676262>.
>



-- 
Sincerely,
Gyu-Ho Lee